src/backend/distributed/utils/shard_utils.c

```diff
@@ -121,3 +124,29 @@ GetLargestShardId()
 
 	return largestShardId;
 }
+
+
+/*
+ * VerifyShardStillExists errors out if the given tasks has any nonexistent
+ * shard id.
+ */
+void
+VerifyShardsStillExist(List *taskList)
+{
+	Task *task = NULL;
+	foreach_ptr(task, taskList)
+	{
+		bool missingOk = true;
+		Oid shardRelationId = LookupShardRelationFromCatalog(task->anchorShardId,
```

Comment:
Hmz, this seems like it might be slow for the common / no split case, since we need to be able to handle hundreds of thousands of queries per second via this path. Can we go via the cache here?